### ISSUE TITLE
Make QuasiQuoters for compile-time parsing

### ISSUE THREAD

**Issue by 3noch** [commercialhaskell/path#59]
In general I prefer QuasiQuoters over TemplateHaskell whenever possible since QQs are less intrusive on syntax and don't require a certain declaration ordering.

**Comment by sjakobi** [comment]
Are there any potential downsides to this apart from the slight API increase? If not, please make a PR!

**Comment by chrisdone** [comment]
> since QQs are less intrusive on syntax and don't require a certain declaration ordering

Can you provide a citation or evidence that an expression quasiquote introduces different ordering behavior to an expression splice?

**Comment by 3noch** [comment]
@chrisdone 

>  Unlike normal declaration splices of the form $(...), declaration quasi-quotes do not cause a declaration group break. See Syntax for more information.

https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#ghc-flag--XQuasiQuotes

TH splices are not able to use declarations that are defined *below* them in the same file. This means that declarations involving TH must have a certain ordering in the file.

**Comment by chrisdone** [comment]
@3noch 

> Unlike normal **declaration** splices of the form $(...), **declaration** quasi-quotes do not cause a declaration group break.

I think that this concerns only declaration splices and declaration quasi-quotes, not expression splices and expression quasi-quotes. The following works:

```haskell
{-# LANGUAGE TemplateHaskell #-}
import Language.Haskell.TH
x = $(varE 'foo)
foo = 123
```

I'm not against `QuasiQuoters` for this feature, but TH is a powerful feature and we should be clear on how it works.

**Comment by 3noch** [comment]
@chrisdone Great! Thanks for the clarification.

**Comment by sjakobi** [comment]
Addressed in #83 and included in v0.5.13.

**Comment by 3noch** [comment]
Thank you @sjakobi!
